sql/schemachanger: add support for unsetting schema_locked inside the legacy schema changer

fqazi opened the pull request:
This PR does the following to enable legacy schema changer support for schema_locked:

1) Adds support for ignoring schema_locked updates that will be ignored by the schema feed. Specifically the schema feed only cares about: `ADD / DROP COLUMN , TRUNCATE, ALTER PRIMARY KEY, ALTER LOCALITY`
3) Update logictest to add a local-schema-locked config. This has been enabled for alter_table, fks, and create_index tests so far.
4) Enable the local-schema-locked by default and update expected files for most tests. Tests using transactions will skip this config for now (~35 tests).

The supported legacy schema changes are following: 
1. **`ALTER TABLE` commands:**
    - `RENAME COLUMN`
    - `RENAME CONSTRAINT`
    - `SET STORAGE_PARAMETER`
    - `INJECT STATISTICS`
    - `RESET STORAGE_PARAMETER`
    - `PARTITION BY` (on the table)
    - `SET ON UPDATE`
    - `DROP NOT NULL`
    - `SET VISIBLE` / `SET NOT VISIBLE`
    - `DROP STORED`
    - `VALIDATE CONSTRAINT`

2. **Other statement types:**
    - `ALTER INDEX`
    - `DROP TABLE`
    - `RENAME COLUMN`
    - `RENAME INDEX`
    - `RENAME TABLE`



Informs: #129694

fqazi committed:
logictest: add local-schema-locked config

Introduce a new config that has schema_locked enabled by default and
update a subset of schema change tests to use this. The follow tests:
alter_table, fks, create_index are now schema_locked enabled.

Informs: #129694
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144589)
<!-- Reviewable:end -->

fqazi committed:
sql: automatically set / unset schema_locked for legacy statements

Previously, schema_locked had to be manually set or unset by users when
making schema changes. We modified the declarative schema changer to
cover some of the more common backfilling schema changes to set / unset
this automatically. Unfortunately, there are a still a number of schema
changes in the legacy schema changer that are not covered. To address
this, this patch adds support for setting / unsetting schema_locked in
implicit transactions with the legacy schema changer.

This patch adds support for a white list of non-backfilling schema
changes, which do not impact schema locked. Updates made by the schema
changes will be ignored by the schema feed, so unsetting schema_locked
is unnecessary for them.

Informs: #129694
Release note: None

fqazi committed:
logictest: enable local-schema-locked for more tests

Previously, create table for schema_locked was set for a small subset of
tests. This patch enables it for all logictests, and excludes ones that
do complex transactional DDL. These will be updated in a later PR.

Informs: #129694

Release note: None

fqazi commented:
@rafiss RFAL

rafiss reviewed (changes requested):
nice work! just had some small questions

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-:-OPl2K6F51hhmuvs-mzT:b-o8rbgd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @michae2)

___
*[`pkg/sql/alter_table.go` line 2358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-OOUlnTPA2HQm_z1t2AK:-OOUlnTPA2HQm_z1t2AL:b-k50d9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6086051dc36d94eb902b5de043a1a55bd49f1bc8/pkg/sql/alter_table.go#L2358)):*
> ```Go
> 	// be simple non-back filling schema changes. All commands except set/reset
> 	// schema_locked are unsupported before 25.2
> 	unsupportedCmd := !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.V25_2) &&
> ```

nit: should this be v25_3?
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-OPl1uZk58wCepQs-yhv:-OPl1uZk58wCepQs-yhw:b-9c8d1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d9503081439a2beb7894b1343e7c014ab7f56ba/pkg/sql/logictest/testdata/logic_test/alter_table#L59)):*
> ```
> LIMIT 1
> ----
> ALTER TABLE test.public.t ADD CONSTRAINT IF NOT EXISTS foo UNIQUE (f)  root  succeeded  1  ·
> ```

do we know why schema_locked has the IF NOT EXISTS part?
___
*[`pkg/sql/alter_table.go` line 2385 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-OPgL9Qm4aPKkYAu_ftF:-OPgL9Qn9ur7KEFZUGQ4:beovw3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1108ad294a4e2fcd13307a1cf26d8db13aef7acb/pkg/sql/alter_table.go#L2385)):*
> ```Go
> 				*tree.AlterTableValidateConstraint, *tree.AlterTableInjectStats:
> 			default:
> 				unsupportedCmd = true
> ```

super nit: let's name the var `preventedBySchemaLocked`


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144589)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao, @michae2, and @rafiss)

___
*[`pkg/sql/alter_table.go` line 2385 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-OPgL9Qm4aPKkYAu_ftF:-OPlFq0L-D3zpP0MsAYk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1108ad294a4e2fcd13307a1cf26d8db13aef7acb/pkg/sql/alter_table.go#L2385)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: let's name the var `preventedBySchemaLocked`
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-OPl1uZk58wCepQs-yhv:-OPlD6a2Bhy0W3rb1oLN:by2bp9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d9503081439a2beb7894b1343e7c014ab7f56ba/pkg/sql/logictest/testdata/logic_test/alter_table#L59)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do we know why schema_locked has the IF NOT EXISTS part?
</blockquote></details>

So, the IF NOT EXISTS works today by the virtue of adding the same elements and not seeing a change. With schema_locked we end up still toggling it let me open an issue for this one as a follow up. 

I think we can add logic inside the builder / planner to collapse things.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144589#-:-OQ4Tp0e1Y0tmUPlKQ6q:b-7m74ox)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @michae2)



<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR!

bors r+
